tests/common/helpers/ptf_tests_helper.py

```diff
@@ -0,0 +1,171 @@
+
+    find_links(duthost, tbinfo, filter)
+    return links
+
+
+@pytest.fixture(scope="module")
+def upstream_links(duthost, tbinfo, nbrhosts):
+    """
+    Returns a dictionary of all the links that are upstream from the DUT.
+
+    Args:
+        duthost: DUT fixture
+        tbinfo: testbed information fixture
+        nbrhosts: neighbor host fixture
+    Returns:
+        links: Dictionary of links upstream from the DUT
+    """
+    links = dict()
+
+    def filter(interface, neighbor, mg_facts, tbinfo):
+        if ((tbinfo["topo"]["type"] == "t0" and "T1" in neighbor["name"])
+                or (tbinfo["topo"]["type"] == "t2" and "T3" in neighbor["name"])):
+            for item in mg_facts["minigraph_bgp"]:
+                if item["name"] == neighbor["name"]:
+                    if isinstance(ip_address(item["addr"]), IPv4Address):
+                        # The address of neighbor device
+                        local_ipv4_addr = item["addr"]
+                        # The address of DUT
+                        peer_ipv4_addr = item["peer_addr"]
+                        break
+            port = mg_facts["minigraph_neighbors"][interface]["port"]
+            links[interface] = {
+                "name": neighbor["name"],
+                "ptf_port_id": mg_facts["minigraph_ptf_indices"][interface],
+                "local_ipv4_addr": local_ipv4_addr,
+                "peer_ipv4_addr": peer_ipv4_addr,
+                "upstream_port": port,
+                "host": nbrhosts[neighbor["name"]]["host"]
+            }
+
+    find_links(duthost, tbinfo, filter)
+    return links
+
+
+def apply_dscp_cfg_setup(duthost, dscp_mode):
+    """
+    Applies the DSCP decap configuration to the DUT.
+
+    Args:
+        duthost: DUT fixture
+        dscp_mode: DSCP mode to apply
+    """
+
+    default_decap_mode = duthost.shell("redis-cli -n 0 hget 'TUNNEL_DECAP_TABLE:IPINIP_TUNNEL' 'dscp_mode'")
+    ["stdout"].strip()
+
+    if default_decap_mode == dscp_mode:
+        return
+
+    for asic_id in duthost.get_frontend_asic_ids():
+        swss = "swss{}".format(asic_id if asic_id is not None else '')
+        cmds = [
+            "docker exec {} cp /usr/share/sonic/templates/ipinip.json.j2 /usr/share/sonic/templates/ipinip.json.j2.tmp"
+            .format(swss),
+            "docker exec {} sed -i 's/{}/{}/g' /usr/share/sonic/templates/ipinip.json.j2 "
+            .format(swss, default_decap_mode, dscp_mode)
+        ]
+        duthost.shell_cmds(cmds=cmds)
+
+    config_reload(duthost, safe_reload=True)
+
+
+def apply_dscp_cfg_teardown(duthost):
+    """
+    Removes the previously applied DSCP decap configuration from the DUT.
+
+    Args:
+        duthost: DUT fixture
+    """
+    for asic_id in duthost.get_frontend_asic_ids():
+        swss = 'swss{}'.format(asic_id if asic_id is not None else '')
+        cmds = [
+            'docker exec {} cp /usr/share/sonic/templates/ipinip.json.j2.tmp /usr/share/sonic/templates/ipinip.json.j2'
```

Comment:
If function `apply_dscp_cfg_setup` didn't run because the expected dscp_mode equals to the current mode, there will be no ipinip.json.j2.tmp. Then this function will throw exception.
Suggest checking the existence of `ipinip.json.j2.tmp` first, and skip the teardown if the file doesn't exist.

Reply:
Good catch